File: 2024/py/src/day04.py
# Day 4: Ceres Search

from src import util

SIZE = 146
EMPTY = "." * SIZE
PAD = "." * 3

path = util.get_input_path("day04.txt")
with open(path, "r") as f:
    ws = [EMPTY, EMPTY, EMPTY]
    for line in f.readlines():
        ws.append(PAD + line.strip() + PAD)

ws.append(EMPTY)
ws.append(EMPTY)
ws.append(EMPTY)


def checkhorz(line, i):
    matches = 0
    if line[i + 1 : i + 4] == "MAS":
        matches += 1
    if line[i - 1 : i - 4 : -1] == "MAS":
        matches += 1

    return matches


def checkvert(li, i):
    matches = 0
    if ws[li + 1][i] == "M" and ws[li + 2][i] == "A" and ws[li + 3][i] == "S":
        matches += 1
    if ws[li - 1][i] == "M" and ws[li - 2][i] == "A" and ws[li - 3][i] == "S":
        matches += 1

    return matches


def checkdiag(li, i):
    matches = 0
    if (
        ws[li + 1][i + 1] == "M"
        and ws[li + 2][i + 2] == "A"
        and ws[li + 3][i + 3] == "S"
    ):
        matches += 1
    if (
        ws[li - 1][i + 1] == "M"
        and ws[li - 2][i + 2] == "A"
        and ws[li - 3][i + 3] == "S"
    ):
        matches += 1
    if (
        ws[li + 1][i - 1] == "M"
        and ws[li + 2][i - 2] == "A"
        and ws[li + 3][i - 3] == "S"
    ):
        matches += 1
    if (
        ws[li - 1][i - 1] == "M"
        and ws[li - 2][i - 2] == "A"
        and ws[li - 3][i - 3] == "S"
    ):
        matches += 1

    return matches


def star1():
    total = 0
    for lidx, line in enumerate(ws):
        for cidx, c in enumerate(line):
            if c != "X":
                continue

            total += checkhorz(line, cidx)
            total += checkvert(lidx, cidx)
            total += checkdiag(lidx, cidx)

    return total


def checkmas(l1, l2, l3):
    return int(
        l2[1] == "A"
        and ((l1[0] == "M" and l3[2] == "S") or (l1[0] == "S" and l3[2] == "M"))
        and ((l1[2] == "M" and l3[0] == "S") or (l1[2] == "S" and l3[0] == "M"))
    )


def star2():
    total = 0
    for i in range(3, 141):
        for j in range(3, 141):
            total += checkmas(
                ws[i][j : j + 3], ws[i + 1][j : j + 3], ws[i + 2][j : j + 3]
            )

    return total

File: 2024/jl/day04.jl
# Day 4: Ceres Search

include("util.jl")

using .Util

ws = Util.getinput("day04.txt")

get(a, i, default='.') = try a[i] catch; default end
getws(x, y, default='.') = try get(ws[x], y) catch; default end

function checkhorz(i, j)
	total = 0
	if getws(i, j+1) == 'M' && getws(i, j+2) == 'A' && getws(i, j+3) == 'S'
		total += 1
	end

	if getws(i, j-1) == 'M' && getws(i, j-2) == 'A' && getws(i, j-3) == 'S'
		total += 1
	end
	
	return total
end

function checkvert(i, j)
	total = 0
	if getws(i+1, j) == 'M' && getws(i+2, j) == 'A' && getws(i+3, j) == 'S'
		total += 1
	end

	if getws(i-1, j) == 'M' && getws(i-2, j) == 'A' && getws(i-3, j) == 'S'
		total += 1
	end
	
	return total
end

function checkdiag(i, j)
	total = 0
	if getws(i+1, j+1) == 'M' && getws(i+2, j+2) == 'A' && getws(i+3, j+3) == 'S'
		total += 1
	end

	if getws(i+1, j-1) == 'M' && getws(i+2, j-2) == 'A' && getws(i+3, j-3) == 'S'
		total += 1
	end

	if getws(i-1, j+1) == 'M' && getws(i-2, j+2) == 'A' && getws(i-3, j+3) == 'S'
		total += 1
	end
	
	if getws(i-1, j-1) == 'M' && getws(i-2, j-2) == 'A' && getws(i-3, j-3) == 'S'
		total += 1
	end

	return total
end

function check(i, j)
	return checkhorz(i, j) + checkvert(i, j) + checkdiag(i, j)
end

function star1()
	total = 0
	for (i, line) ∈ enumerate(ws)
		for (j, c) ∈ enumerate(line)
			if c != 'X'
				continue
			end

			total += check(i, j)
		end
	end

	println("- $total")
end

checkmas(l1, l2, l3) = Int(get(l2, 1) == 'A' && (get((l1, 0) == 'M' && get(l3, 2) == 'S') or (get(l1, 0) == 'S' && get(l3, 2) == 'M')) && ((l1[2] == 'M' && l3[0] == 'S') or (l1[2] == 'S' && l3[0] == 'M')))

function star2()
	total = 0
	for i ∈ eachindex(ws)
		for j ∈ eachindex(ws[i])
			total += checkmas(ws[i][j:j+3],ws[i+1][j:j+3],ws[i+2][j:j+3])
		end
	end

	println("- $total")
end

star1()
star2()


